### DISCUSSION TITLE
Documentation for ConfigurationSchema.json

### DISCUSSION THREAD

**Discussion by Falco20019** [microsoft/aspire#963] [Q&A] [answered]
I was trying to find out how the JSON schema in .NET Aspire components are linked toolwise. Sadly I did not find anything on https://learn.microsoft.com/de-de/dotnet/aspire/ or https://github.com/dotnet/aspire/blob/main/src/Components/README.md.

After checking through GitHub, I recognized that the schemes are defined as 
[ConfigurationSchema.json](https://github.com/dotnet/aspire/blob/main/src/Components/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL/ConfigurationSchema.json). I also did find the https://github.com/dotnet/aspire/blob/main/src/Components/Directory.Build.targets file. I therefore assume it's linked to `ConfigurationSchemaPath` property at build time and and the json file being packaged into the NuGet package together with the `JsonSchemaSegment` item being set as transitive build dependency.

Is Visual Studio just checking for these files or how is everything linked together?

**Comment by DamianEdwards** [comment, marked as answer]
Yes, VS is bringing in project targets that collect up the `JsonSchemaSegment` items that are added by the [targets imported by the component packages](https://github.com/dotnet/aspire/blob/eceed4d2bd168d5a969d90dd5608e213edd345d2/src/Components/Common/Package.targets#L3C6-L3C24). This is then processed when the *appsettings.json* file is opened and the schema is dynamically generated into a location under the the `obj` directory and the editor is pointed at that file as the schema.

Our intent is to move this logic out of VS and into the base .NET SDK so that this mechanism can be used by any NuGet package to dynamically contribute JSON schema to related files within the projects the package is added to. The actual schema generation still needs to be triggered by some gesture, e.g. a JSON file being opened in the IDE, a target being run at the command line, etc.

You should be able to point any JSON schema supporting editor at these generated files by way of the `"$schema"` property in the open JSON file, but as I said, right now the targets that do the actual generation are only shipped in VS.

**Reply by KalleOlaviNiemitalo** [reply]
Which versions of Visual Studio recognize the `JsonSchemaSegment` items?  It seems Visual Studio 15.9.58 doesn't.

**Reply by DamianEdwards** [reply]
Latest preview version (so 17.9.x) and likely only in web projects. @vijayrkn

**Reply by KalleOlaviNiemitalo** [reply]
> Our intent is to move this logic out of VS and into the base .NET SDK

Is there a public issue for tracking the progress of that feature?

We have a monorepo that includes a few logging providers with JSON schemas for configuration.  Currently, each application references the logging projects via `ProjectReference` (not `PackageReference`) and references parts of their configuration schemas via `"$ref"`; but relative URIs in `"$ref"` are somewhat difficult to manage so that they are valid in the source tree and remain valid when the configuration schemas are deployed.  It would be great if .NET SDK were able to merge the configuration schemas at build time and we could ship the merged file in the application package.
